modules/system-source/system-source.c

```diff
@@ -216,46 +216,21 @@ system_sysblock_add_linux_kmsg(GString *sysblock)
                              "kernel", "kernel", format, TRUE);
 }
 
-static gboolean
-_is_running_in_linux_container(void)
-{
-  FILE *f;
-  char line[2048];
-  gboolean container = FALSE;
-
-  f = fopen("/proc/1/cgroup", "r");
-  if (!f)
-    return FALSE;
-
-  while (fgets(line, sizeof(line), f) != NULL)
-    {
-      if (line[strlen(line) - 2] != '/')
-        {
-          container = TRUE;
-          break;
-        }
-    }
-
-  fclose (f);
-
-  return container;
-}
-
 static void
-system_sysblock_add_linux(GString *sysblock)
+system_sysblock_add_linux(GString *sysblock, CfgArgs *args)
 {
   if (service_management_get_type() == SMT_SYSTEMD)
     system_sysblock_add_systemd_source(sysblock);
   else
     {
       system_sysblock_add_unix_dgram(sysblock, "/dev/log", NULL, "8192");
-      if (!_is_running_in_linux_container())
+      if (cfg_args_get(args, "exclude_kmsg") == NULL)
```

Comment:
This only checks that the option is present, and does not account for things like `exclude-kmsg(no)`. If we require the presence of the option only (which is a great idea, by the way!), we should either mandate that it has no arguments, or treat at least some falsey values as if the option wasn't present.

For simplicity's sake, I'd do something like this:

```c
if (cfg_args_contains(args, "exclude_kmsg")) {
  if (cfg_args_get(args, "exclude_kmsg")) {
    msg_error("system(): The exclude_kmsg() option takes no arguments.");
    return FALSE;
  }
  system_sysblock_add_linux_kmsg(sysblock);
}
```

This would require `system_sysblock_add_linux` to start returning `gboolean`, and the other branches should return `TRUE`. We'd also need to adjust `system_generate_system_transports` below to check the return value and do the right thing (`goto exit`).

We need two steps (`_contains` and `_get`), because we can't simply check if the value of the argument is empty, because that would be true even if the option wasn't present. So we first make sure it is present, and only THEN check the value.

Reply:
Thank You @algernon for taking time on this. I really like your idea, which adds more flexibility to the feature. First I was overwhelmed with the concept to not strict this settings any more than needed, just silently accept any value, so it can be expanded - without syntax modification - in the future.

In the meantime I discussed this question IRL with @mitzkia. To get some "best practice" from testers point of view.

New approach: introduce the **flags** keyword to the system source, and add **exclude-kmsg** as the first possible item.

**pros**
- **flags** is already a well known keyword, so it will align well with the current logic.
- It will represent the "present or not" nature of this option.

**cons**
- It will be a little bit confusing since the namespace will be different than: source drivers, syslog parser, log path's. (Flags are different for these items.)

note: We are trying to generalize a single option. It is very good to "try to think about the future", but I think it is almost impossible. So I will gladly accept any new opinion in the topic, and ready to implement the most popular one.

Reply:
While I like the `flags` idea too, in context of `system-source`, it isn't as simple to implement as `exclude-kmsg()` would have been, because it would need to break up the string into list items... on the other hand, since we only have one flag now, perhaps that's not needed, and a simple `strcmp(__normalize_key(flags), "exclude_kmsg")` would do?

In other words, I'll back the `flags(exclude-kmsg)` idea.